### PR TITLE
Significantly extend README.md

### PR DESCRIPTION
**Pull request by david-a-wheeler**
The original README.md said almost nothing.
Let's modify it so the README.md actually provides more detail.

I tried to capture the rule for approving merges per the list,
but with more precision. If you think I got it wrong, please
tell me how to fix it! I think it's important to document the
rule(s) for merging changes.

Signed-off-by: David A. Wheeler <dwheeler@dwheeler.com>

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,7 +2,191 @@
...
+* checkmm (a C++ verifier by Eric Schmidt)
+* smetamath-rs (smm3) (a Rust verifier by Stefan O'Rear)
+* mmj2 (a Java verifier by Mel L. O'Cat and Mario Carneiro)
+* mmverify.py (a Python verifier by Raph Levien)
+
+Note that these are different verifiers written in different programming
+languages by different people. In addition, the verification algorithm
+is intentionally simple (it fits in two pages in the Matamath book).
+This provides *extremely* high confidence that the proofs are correct.
+Most math proofs are not formally verified at all (that is, where a
```
Comment (tirix) [inline README.md]: Could we make it clear that this does _not_ refer to Metamath proofs?

File: README.md [hunk 2]
```diff
@@ -2,7 +2,191 @@
...
+[video "Metamath Proof Explorer: A Modern Principia Mathematica](https://www.youtube.com/watch?v=8WH4Rd4UKGE).
+
+## What databases are included in this collection?
+
+This is a collection of many databases, but since "set.mm" is by far
+the most actively maintained, you may want to look at it first even if you
+plan to eventually work on something else.
+[The video "Metamath Proof Explorer (set.mm) contributions visualized with Gource through 2019-10-04"]( https://www.youtube.com/watch?v=XC1g8FmFcUU&list=PL1jSu6GGefBm7RBP0Id2Sa9uyVuyhioAC&index=4) shows set.mm's growth over time.
+
+The databases included, in (approximate) decreasing size, are:
```
Comment (tirix) [inline README.md]: I think there are a few other Metamath databases outside of this repository. It would be interesting to try to list them all somewhere, and point to that list from here.

File: README.md [hunk 3]
```diff
@@ -2,7 +2,191 @@
...
+
+Metamath is a computer language and associated computer program for
+archiving, verifying, and studying mathematical proofs.
+
+Unlike some other systems, Metamath does not build a particular set
+of axioms into the system. Instead, the Matamath language is simple and robust,
+with an almost total absence of hard-wired syntax.
+In Metamath you express the axioms *and* the theorems in a database.
+To prove a theorem, every proof step *must* be proven using an axiom or
+a previously-proven theorem; as a result, nothing is hidden.
```
Comment (bprzybocki) [inline README.md]: no hyphen after -ly

File: README.md [hunk 4]
```diff
@@ -2,7 +2,191 @@
...
+We believe Metamath provides about the simplest possible framework that
+allows essentially of mathematics to be expressed with absolute rigor.
```
Comment (bprzybocki) [inline README.md]: essentially *all*?

File: README.md [hunk 5]
```diff
@@ -2,7 +2,191 @@
...
+
+For more information see
+the [Metamath Home Page](http://us.metamath.org/), the
+[Metamath Proof Explorer Home Page](http://us.metamath.org/mpeuni/mmset.html),
+the [Metamath book](http://us.metamath.org#book), or the
+[video "Metamath Proof Explorer: A Modern Principia Mathematica](https://www.youtube.com/watch?v=8WH4Rd4UKGE).
```
Comment (bprzybocki) [inline README.md]: missing close quote

File: README.md [hunk 6]
```diff
@@ -2,7 +2,191 @@
...
+
+## What databases are included in this collection?
+
+This is a collection of many databases, but since "set.mm" is by far
+the most actively maintained, you may want to look at it first even if you
+plan to eventually work on something else.
+[The video "Metamath Proof Explorer (set.mm) contributions visualized with Gource through 2019-10-04"]( https://www.youtube.com/watch?v=XC1g8FmFcUU&list=PL1jSu6GGefBm7RBP0Id2Sa9uyVuyhioAC&index=4) shows set.mm's growth over time.
```
Comment (bprzybocki) [inline README.md]: Isn't there a more recent version (2020) of this video?

File: README.md [hunk 7]
```diff
@@ -2,7 +2,191 @@
...
+
+The databases included, in (approximate) decreasing size, are:
+
+* "set.mm" aka Metamath Proof Explorer (MPE) -
+  uses classical logic and
+  Zermelo–Fraenkel set theory with the axiom of choice (ZFC).
+* "iset.mm" aka "Intuitionistic Logic Explorer" -
+  uses intuitionistic logic instead of classical logic,
+  e.g., it does not presume that the law of the implied middle is true
```
Comment (bprzybocki) [inline README.md]: I think you mean "excluded", unless "implied middle" is an alternative terminology I've never heard of.

File: README.md [hunk 8]
```diff
@@ -2,7 +2,191 @@
...
+  in all cases.
```
Comment (bprzybocki) [inline README.md]: You may want to note that iset.mm is intuitionistic *set theory*, because there are several other varieties of intuitionistic mathematics. Specifically, I think iset.mm uses IZF (which is strictly stronger than CZF, another version of intuisionitic set theory).

File: README.md [hunk 9]
```diff
@@ -2,7 +2,191 @@
...
+## What databases are included in this collection?
+
+This is a collection of many databases, but since "set.mm" is by far
+the most actively maintained, you may want to look at it first even if you
+plan to eventually work on something else.
+[The video "Metamath Proof Explorer (set.mm) contributions visualized with Gource through 2019-10-04"]( https://www.youtube.com/watch?v=XC1g8FmFcUU&list=PL1jSu6GGefBm7RBP0Id2Sa9uyVuyhioAC&index=4) shows set.mm's growth over time.
+
+The databases included, in (approximate) decreasing size, are:
+
+* "set.mm" aka Metamath Proof Explorer (MPE) -
```
Comment (bprzybocki) [inline README.md]: To be consistent with the other entries, "Metamath Proof Explorer" should be in quotes.

**Review by sctfn — APPROVED** [review]

File: README.md [hunk 10]
```diff
@@ -2,7 +2,191 @@
...
+  uses classical logic and
+  Zermelo–Fraenkel set theory with the axiom of choice (ZFC).
+* "iset.mm" aka "Intuitionistic Logic Explorer" -
+  uses intuitionistic logic instead of classical logic,
+  e.g., it does not presume that the law of the implied middle is true
+  in all cases.
+* "nf.mm" aka "New Foundations Explorer" - constructs mathematics using
+  Quine's "New Foundations" (NF) set theory axioms, a direct derivative
```
Comment (bprzybocki) [inline README.md]: I don't see any reason for "New Foundations" to be in quotes.

File: README.md [hunk 11]
```diff
@@ -2,7 +2,191 @@
...
+* "ql.mm" aka "Quantum Logic Explorer" - Starts from the orthomodular
+  lattice properties proved in the Hilbert Space Explorer and takes
+  you into quantum logic.
+* "hol.mm" aka "Higher-Order Logic Explorer" - Starts with HOL (also
+  called simple type theory) and derives equivalents to ZFC axioms,
+  connecting the two approaches.
+* "peano.mm" - Peano arithmetic.
+* "big-unifier.mm" - a unification and substitution test for
+  Metamath verifiers, where small input expressions blow up to thousands
+  of symbols.  It is a translation of William McCune's "big-unifier.in"
```
Comment (bprzybocki) [inline README.md]: Two spaces after the period is inconsistent with the rest of the document

File: README.md [hunk 12]
```diff
@@ -2,7 +2,191 @@
...
+* "demo0.mm" - a simple formal system used as a demonstration in
+  Chapter 2 of the Metamath book.
+
+## How are they verified?
+
+Changes ("commits") to any database are first automatically verified
+before they are accepted, using GitHub actions.
+
+The set.mm and iset.mm databases are the ones primarily being updated.
+Every change to these databases are verified by the following
```
Comment (bprzybocki) [inline README.md]: "are" -> "is"

File: README.md [hunk 13]
```diff
@@ -2,7 +2,191 @@
...
+* smetamath-rs (smm3) (a Rust verifier by Stefan O'Rear)
+* mmj2 (a Java verifier by Mel L. O'Cat and Mario Carneiro)
+* mmverify.py (a Python verifier by Raph Levien)
+
+Note that these are different verifiers written in different programming
+languages by different people. In addition, the verification algorithm
+is intentionally simple (it fits in two pages in the Matamath book).
+This provides *extremely* high confidence that the proofs are correct.
+Most math proofs are not formally verified at all (that is, where a
+computer verification verifies every step). The rest are typically only
```
Comment (bprzybocki) [inline README.md]: "verification" should probably be deleted here

File: README.md [hunk 14]
```diff
@@ -2,7 +2,191 @@
...
+verified by a single program (which might have an error in it).
+These multiple independent verifications provide very high confidence
+of their being correct.
```
Comment (bprzybocki) [inline README.md]: "their" should probably be "the proofs'" (it's more natural to say a proof is correct than to say a verification is correct, IMO)

File: README.md [hunk 15]
```diff
@@ -2,7 +2,191 @@
...
+[Metamath Proof Explorer (MPE, aka set.mm)](http://us.metamath.org/mpegif/mmset.html) is a good way to start.
+
+To create a proof, you need to pick a tool.
+The Metamath system specifies a common file format, not a single tool
+everyone has to use.
+Many in the community use the [mmj2 tool](http://us.metamath.org#book).
+The video [Introduction to Metamath and mmj2](https://www.youtube.com/watch?v=Rst2hZpWUbU) demonstrates using mmj2.
+The mmj2 tool includes an interactive tutorial, and if you just want to watch,
+there's a video giving a
+[Walkthrough of the tutorial in mmj2](https://www.youtube.com/watch?v=87mnU1ckbI0&t=2094s).
```
Comment (bprzybocki) [inline README.md]: Do you mean to link to the video starting at a timestamp halfway through the video?

File: README.md [hunk 16]
```diff
@@ -2,7 +2,191 @@
...
+You should start small and try to reprove what's already
+proven with Metamath before you try to prove something new.
+If you're contributing to set.mm or iset.mm, take a look at its
+[conventions and style](http://us.metamath.org/mpegif/conventions.html).
+
+At some point you should join the
+[Metamath mailing list](https://groups.google.com/g/metamath).
+We're a friendly community and we would love to have more collaborators!
```
Comment (bprzybocki) [inline README.md]: comma before and

File: README.md [hunk 17]
```diff
@@ -2,7 +2,191 @@
...
+blemish like an odd name. It's a pain to fix "merge conflicts"
+because differences have accumulated over time, and we'd rather avoid
+merge conflicts if there's no good reason to have them.
+
+We encourage changes to be smaller, focusing on a single logical idea.
+That makes changes much easier to review.
+It also makes it easy to accept some changes while not accepting others.
+
+Many contributors have or create a personal "mathbox" - a sandbox of work that
+is visible to others. Changes to ones' own mathbox still go through a
```
Comment (bprzybocki) [inline README.md]: "ones'" -> "one's"

File: README.md (hunk 8 above)
Comment (jkingdon) [reply README.md]: There is some CZF starting with the section "Bounded formulas" just above http://us.metamath.org/ileuni/ax-bd0.html . Most of iset.mm is IZF, though.

And of course it is true that this is all set theory, not type theory or another style of foundation.

How deep to get into this in the README, I'm not quite as clear, but I'm all for explaining these distinctions, whether here or elsewhere.

File: README.md [hunk 18]
```diff
@@ -2,7 +2,191 @@
...
+We're a friendly community and we would love to have more collaborators!
+
+The Wiki page
+[Getting started with contributing](https://github.com/metamath/set.mm/wiki/Getting-started-with-contributing) will walk you the process of contributing
+a final proof once you have one.
+Basically, you'll create a "pull request" on GitHub to request
+merging of some change(s).
+**Warning**: Changes are made to the **develop** branch.
+
+## What are the requirements for acceptance (merging)?
```
Comment (jkingdon) [inline README.md]: I think this section should be moved to the https://github.com/metamath/set.mm/blob/develop/CONTRIBUTING.md file.

The README should say it is great to contribute and something about how to write proofs and the like, but the details of pull requests and that sort of thing belong in CONTRIBUTING if I correctly understand the usual github conventions.

File: README.md [hunk 19]
```diff
@@ -2,7 +2,191 @@
...
+Similarly, mathboxes should normally only be changed by the owner of the
+mathbox (unless it's a general improvement that applies across the database).
+We don't currently have a list of area maintainers; please propose
+a change to this README (here) if you want to be an area maintainer.
+
+## Credits and a memorium
+
+Our sincere thanks to *everyone* who has contributed to this work.
+
+In addition, this entire collection is dedicated to the memory of
```
Comment (jkingdon) [inline README.md]: Well done. We can also think about writing something at the start of Norm's mathbox (similar to the one we have for Alan Sare), but this doesn't have to be done all at once.

File: README.md (hunk 7 above)
Comment (jkingdon) [reply README.md]: "law of the excluded middle" is pretty much universal in recent works as far as I've noticed (I'm including variant wordings like  "excluded middle" or "LEM").

**Review by jkingdon — COMMENTED** [review]
Thanks for this expansion; although I'm maybe a tiny bit worried about having similar text here and in http://us.metamath.org/ and in the book and <etc>, on balance I agree it is better to have intros each place (probably slightly different in emphasis) rather than trying to have one perfect one which all the others point to.

File: README.md (hunk 3 above)
Comment (david-a-wheeler) [reply README.md]: I think either way is fine, but I'll remove the hyphen just to make sure :-).

File: README.md (hunk 4 above)
Comment (david-a-wheeler) [reply README.md]: Yes!

File: README.md (hunk 5 above)
Comment (david-a-wheeler) [reply README.md]: Agree.

File: README.md (hunk 6 above)
Comment (david-a-wheeler) [reply README.md]: Yes, and I'm the one who created it :-). Fixed. Thanks!

File: README.md (hunk 9 above)
Comment (david-a-wheeler) [reply README.md]: Agre.

File: README.md (hunk 7 above)
Comment (david-a-wheeler) [reply README.md]: I have no idea how that happened. Agreed!!!

File: README.md (hunk 8 above)
Comment (david-a-wheeler) [reply README.md]: I don't think we need to get into IZF vs. CZF here. If someone's interested, they'll jump to that.

Revised summary;

* "iset.mm" aka "Intuitionistic Logic Explorer" -  uses intuitionistic set theory.  In particular, it does not presume that the law of excluded middle is  necessarily true in all cases.

File: README.md (hunk 11 above)
Comment (david-a-wheeler) [reply README.md]: It won't matter since this is markdown that generates HTML. But sure, I'll change it.

File: README.md (hunk 12 above)
Comment (david-a-wheeler) [reply README.md]: Agreed. I'll tweak the wording further so it's clear what's happening.

File: README.md (hunk 13 above)
Comment (david-a-wheeler) [reply README.md]: Agreed.

File: README.md (hunk 14 above)
Comment (david-a-wheeler) [reply README.md]: Agreed. I'll add a few clarifying words to bring the point home.

File: README.md (hunk 15 above)
Comment (david-a-wheeler) [reply README.md]: You're right, thanks for noticing! Fixed in the version I'm about to post.

File: README.md (hunk 16 above)
Comment (david-a-wheeler) [reply README.md]: the comma isn't needed, but happy to add it.

File: README.md (hunk 18 above)
Comment (david-a-wheeler) [reply README.md]: That is the usual convention. These instructions are pretty short, so I thought I could get away with just slipping them into the README. Happy to move them to a CONTRIBUTING.md file, though - but if I do, I'll create a link to it from the README.

File: README.md (hunk 17 above)
Comment (david-a-wheeler) [reply README.md]: I think it's clearer if we say "someone's", so I'll do that instead.

File: README.md (hunk 19 above)
Comment (david-a-wheeler) [reply README.md]: Thanks! Since Norm created all this, I think it's reasonable as a special case to put it here in the README. But you're right, it'd be good to say something in his mathbox too... but let's do that separately. In fact, if this is in the README, I think there's no hurry.

File: README.md (hunk 18 above)
Comment (david-a-wheeler) [reply README.md]: Done.

File: README.md (hunk 10 above)
Comment (david-a-wheeler) [reply README.md]: Agreed.

**Comment by david-a-wheeler** [comment]
Okay, I think I responded to all proposed changes, including moving some of the material into CONTRIBUTING.md. Any other comments?

**Review by digama0 — APPROVED** [review]

**Comment by savask** [comment]
I have some nitpicks which mostly apply to style, so others may disagree.

**First and second sections**
My general feeling is that the text is too wordy at some places, so my suggestions are mostly for removing things :-) For example, the first sentence should express the main goal of the repository shortly and clearly, something like "This is a collection of mathematical databases written in the Metamath proof language." Certainly there's no need to explain what a "database" is, or mention the international community or 5 verifiers.

The first paragraph (or even the first sentence) should contain a link to the main metamath website - right now we have a link to the 100 theorems list, but actual useful metamath links can be found only at the end of the second section.

**Third section**
The link to gource is, in my opinion, not relevant here, one can see that set.mm is the largest file without it.

**Fourth section**
I already mentioned that in my opinion one should remove the sentence about 5 verifiers from the first paragraph (since it may mislead people into thinking that metamath databases must be verified by a suite of 5 programs or something), and I think one should remove an explicit list of 5 verifiers from the fourth section as well or, at least, rewrite it. I know that @david-a-wheeler highly values the fact that metamath databases can be verified by independent programs, but not everyone values that equally and it may confuse a newcomer. Instead I suggest writing that databases are verified by 5 independent programs (without elaborating the philosophical implications of that), and explaining that additionally markup is checked by metamath-exe, and definitions are checked by mmj2.

**Comment by digama0** [comment]
I generally agree with @savask . My first impression is that the readme is too wordy and not goal-directed enough. People read a readme to find out what it is and how to use it, they don't need a history lesson. (It fits in the "how-to guide" part of the [classification of kinds of documentation](https://documentation.divio.com/).) That said, I don't have too many concrete wording suggestions.

File: CONTRIBUTING.md [hunk 20]
```diff
@@ -1,109 +1,93 @@
-If you are new to GitHub and git, the basic instructions for
```
Comment (jkingdon) [inline CONTRIBUTING.md:1]: Are you intentionally getting rid of the entire former contents of the CONTRIBUTING.md file? Some parts of it, like the ones about rewrapping and the discouraged file, seem useful for sure. Others are perhaps best kept unless we decide otherwise?

**Comment by jkingdon** [comment]
> My first impression is that the readme is too wordy and not goal-directed enough. People read a readme to find out what it is and how to use it, they don't need a history lesson.

I guess probably I agree with this but quite aside from philosophy, I think the specific suggestions that savask makes are sound ones, so perhaps we can make those edits (or for some items decide not to if @david-a-wheeler feels strongly about that content being in this particular location). Overall questions of purpose and scope are good ones to keep in mind for future revisions but I don't see any particular need we have to shoot for perfection all at once.

**Review by jkingdon — CHANGES_REQUESTED** [review]

File: CONTRIBUTING.md [hunk 21]
```diff
@@ -1,109 +1,93 @@
...
+[Metamath Proof Explorer (MPE, aka set.mm)](http://us.metamath.org/mpegif/mmset.html) is a good way to start.
+
+To create a proof, you need to pick a tool.
+The Metamath system specifies a common file format, not a single tool
+everyone has to use.
+Many in the community use the [mmj2 tool](http://us.metamath.org#book).
+The video [Introduction to Metamath and mmj2](https://www.youtube.com/watch?v=Rst2hZpWUbU) demonstrates using mmj2.
+The mmj2 tool includes an interactive tutorial, and if you just want to watch,
+there's a video giving a
+[Walkthrough of the tutorial in mmj2](https://www.youtube.com/watch?v=87mnU1ckbI0).
```
Comment (avekens) [inline CONTRIBUTING.md:34]: I think metamath.exe should be mentioned here, too. For me, it was more adeqate to start with this tool to become familiar with the essentials of Metamath. And I still use it for most of my work (and I think Norm did it, too).

File: CONTRIBUTING.md [hunk 22]
```diff
@@ -1,109 +1,93 @@
...
+At some point you should join the
+[Metamath mailing list](https://groups.google.com/g/metamath).
+
+The Wiki page
+[Getting started with contributing](https://github.com/metamath/set.mm/wiki/Getting-started-with-contributing) will walk you the process of contributing
+a final proof once you have one.
+Basically, you'll create a "pull request" on GitHub to request
+merging of some change(s).
+**Warning**: Changes are made to the **develop** branch.
+
```
Comment (avekens) [inline CONTRIBUTING.md:54]: I would change the order of these sentences: First say where to contribute (GitHub), and then how. 
Suggestion: 
"For contributing a final proof once you have one, you'll basically create a "pull request" on GitHub to request
merging of some change(s) into the corresponding Git repository.
**Warning**: Changes are made to the **develop** branch.

The Wiki page [Getting started with contributing](https://github.com/metamath/set.mm/wiki/Getting-started-with-contributing) will walk you the process of setting up the GitHub environment and contributing your proof."

File: README.md [hunk 23]
```diff
@@ -2,7 +2,127 @@
 
 # Metamath set.mm repository
 
-This is a set of Metamath source files (databases) for logic and set theory.
+This is a collection of databases that specify mathematical axioms and
+rigorous formal proofs of various mathematical theorems
+derived from those axioms.
+This collection is specified and derived using the Metamath system.
+A "database" in Metamath parlance is a collection of one or more text files.
```
Comment (avekens) [inline README.md]: Maybe to point to the *.mm-files which "are" the databases" would be usefull, because there are many other files in this repository.

File: README.md [hunk 24]
```diff
@@ -2,7 +2,127 @@
...
-File "iset.mm" uses intuitionistic logic.
+This collection is being actively improved by an international community.
+Please join us!
+
+The two largest databases in this collection are "set.mm" and "iset.mm".
+The "set.mm" database is one of the largest
+collections of formally-verified mathematics in the world (e.g., it
+has completed many challenge theorems in the
+[Formalizing-100 Theorems](https://www.cs.ru.nl/~freek/100/) challenge list).
+It is also one of the most rigorously verified collection of
```
Comment (avekens) [inline README.md]: As already suggested by others, references to http://us.metamath.org/mpeuni/mmset.html (or http://us.metamath.org/mpeuni/mmtheorems.html?) and http://us.metamath.org/ileuni/mmil.html (or http://us.metamath.org/ileuni/mmtheorems.html?) should be added here.

File: CONTRIBUTING.md (hunk 20 above)
Comment (avekens) [reply CONTRIBUTING.md:1]: I agreee with @jkingdon : This deleted material should be kept. Maybe at/near the end of the page?

**Review by avekens — CHANGES_REQUESTED** [review]
Great work, @david-a-wheeler ! I have only minor remarks (see inline comments) which you may consider.

**Review by avekens — CHANGES_REQUESTED** [review]
I requested some changes... ?

**Comment by avekens** [comment]
I have performed a review and requested some changes:
grafik
It seems, however, that this was not handled by Github as real change request, as done for @jkingdon . Maybe I do not have sufficient permissions to do it?

**Comment by benjub** [comment]
The material currently in the CONTRIBUTING.md file is very useful, please do not remove it.

I agree that the proposal is a bit wordy too; some detail/history can be removed.
Proposals:
Matamath --> Metamath (twice)
formally-verified --> formally verified (twice)
We-generally --> We generally
5 --> five

In the last paragraph, I would remove "In addition".

File: CONTRIBUTING.md (hunk 20 above)
Comment (david-a-wheeler) [reply CONTRIBUTING.md:1]: Sorry, that's my mistake. It was suggested that I move the contribution information to CONTRIBUTING.md, so I did that in a revision to this commit, but in the process I accidentally deleted what was there!

Good news: Since we're using a version control system, it's easy to get that information. I'll try to merge the old/new material into a hopefully-consistent whole.

File: README.md (hunk 23 above)
Comment (david-a-wheeler) [reply README.md]: Fair enough. I'll add hyperlinks from each database name in the list of databases. That's probably the easiest way to do that.

File: README.md (hunk 24 above)
Comment (david-a-wheeler) [reply README.md]: Agreed!

I think the right starting points are http://us.metamath.org/mpeuni/mmset.html  and http://us.metamath.org/ileuni/mmil.html - they provide context & then link on to their table of contents (as well as providing other useful starting points.

File: CONTRIBUTING.md (hunk 20 above)
Comment (david-a-wheeler) [reply CONTRIBUTING.md:1]: Okay, I've merged the old and new material into something that is hopefully coherent.

File: CONTRIBUTING.md (hunk 21 above)
Comment (david-a-wheeler) [reply CONTRIBUTING.md:34]: Will do.

File: CONTRIBUTING.md (hunk 22 above)
Comment (david-a-wheeler) [reply CONTRIBUTING.md:54]: That makes sense! Done.

File: CONTRIBUTING.md (hunk 21 above)
Comment (david-a-wheeler) [reply CONTRIBUTING.md:34]: Will do.

File: CONTRIBUTING.md (hunk 22 above)
Comment (david-a-wheeler) [reply CONTRIBUTING.md:54]: Done.

**Comment by david-a-wheeler** [comment]
@savask -

How about this, why don't we greatly shorten the *early* part, but keep the discussion about multiple verifiers later on? I think the multiple verifiers is important, in particular because it's *distinctive*, but we can place that information later where people who don't care can easily skip it (while people who *do* care can easily see it).

New section 1:

> This is a collection of rigorously-verified [Metamath](http://us.metamath.org/) databases that specify mathematical axioms and formal proofs of theorems derived from those axioms.

> Certainly there's no need to explain what a "database" is, or mention the international community or 5 verifiers.

I think it does need to be explained somewhere, because the word "database" used by Metamath isn't really the same meaning as used by others. It's similar but not the same. But we can move the definition down to where the actual list of databases is discussed; people don't need to know the difference until then.

> The first paragraph (or even the first sentence) should contain a link to the main metamath website - right now we have a link to the 100 theorems list, but actual useful metamath links can be found only at the end of the second section.

Done.

> The link to gource is, in my opinion, not relevant here, one can see that set.mm is the largest file without it.

I'll slip it into the set.mm discussion, where it makes more sense anyway.

> I already mentioned that in my opinion one should remove the sentence about 5 verifiers from the first paragraph (since it may mislead people into thinking that metamath databases must be verified by a suite of 5 programs or something), and I think one should remove an explicit list of 5 verifiers from the fourth section as well or, at least, rewrite it. I know that @david-a-wheeler highly values the fact that metamath databases can be verified by independent programs, but not everyone values that equally and it may confuse a newcomer. Instead I suggest writing that databases are verified by 5 independent programs (without elaborating the philosophical implications of that), and explaining that additionally markup is checked by metamath-exe, and definitions are checked by mmj2.

I'll remove that material in section 1, and put it all in one place. Hopefully that'll be "good enough" for this first cut at a README update.

**Review by jkingdon — APPROVED** [review]

**Comment by david-a-wheeler** [comment]
Okay, I just made a bunch of changes. In almost all cases I made the specific changes requested. In a few I didn't (e.g., I kept the list of 5 checkers because I think it's important), but in those cases I greatly shortened the README's first section (which I think was the *main* concern) & so hopefully in those cases this is a "reasonable compromise".

I'm sure we'll make many more changes, this isn't the be-all end-all. It's just an attempt to make the README provide some actual useful information, as well as documenting our new process for approving merges (since Norm is no longer here to make those calls).

**Comment by david-a-wheeler** [comment]
I think I did *something* in response to every comment. But if you think I missed something, or strongly disagree with the resulting text in some way, please let me know!!

**Review by tirix — APPROVED** [review]

File: README.md (hunk 2 above)
Comment (david-a-wheeler) [reply README.md]: Interesting, yes. I propose making that a separate pull request. And ideally (for me), you'd be making it :-).

File: README.md (hunk 1 above)
Comment (david-a-wheeler) [reply README.md]: Okay. I just tweaked the text, I hope it does that.

**Review by tirix — APPROVED** [review]

File: CONTRIBUTING.md (hunk 21 above)
Comment (avekens) [reply CONTRIBUTING.md:34]: not done yet - in a subsequent PR?

**Review by avekens — APPROVED** [review]

**Comment by savask** [comment]
@david-a-wheeler 
Let me comment on some other changes, and then I'll rant a little bit about 5 verifiers again.

In the **first** section I would remove the phrase "Please join us to improve it further!", since it is a bit weird to ask for contributions before sections "What is Metamath?" and "How can I contribute?" :-) The penultimate section also asks for contributions, so a similar phrase can be removed from the first section, in my opinion.

In the **third** section I suggest removing everything before the phrase "The databases included and links to their generated displays, in (approximate) decreasing size, are:". The fact that a "database" is some sort of file is pretty obvious and becomes even clearer once the reader sees links to files in the database list.

Also maybe I'm reading an older version, but I still see two occurrences of "M**a**tamath" in the text.

Now about multiple verifiers.
>How about this, why don't we greatly shorten the early part, but keep the discussion about multiple verifiers later on? I think the multiple verifiers is important, in particular because it's distinctive, but we can place that information later where people who don't care can easily skip it (while people who do care can easily see it).

My point is that not everyone values there being multiple verifiers, and having this information in the README puts an unusually high emphasis on the matter. This topic has been discussed multiple times earlier, see for example issues #1202 and especially #1867 , and it's easy to see that there are quire polar opinions on that. I merely suggest to avoid this controversial topic in the README and mention only the bare minimum, something like:

>Changes ("commits") to any database are first automatically verified before they are accepted, using GitHub actions. Every change to set.mm and iset.mm is verified by several independent verifiers, most notably [metamath-exe](https://github.com/metamath/metamath-exe), which also checks markup, and [mmj2](https://github.com/digama0/mmj2), which checks definition [soundness](http://us.metamath.org/mpeuni/mmset.html#definitions). All other databases' proofs are verified by one verifier (metamath-exe).

And that's it, without listing all the other verifiers or explaining the underlying reason behind all this.

I hope this doesn't come off as a personal grudge against multiple verifiers (because it's not), but my main message is that Metamath community members may have different views on things and disputable places should be avoided. Some, like Norman, value human-readabilty of set.mm and it being minimized almost to the byte-count, some like its focus on ZFC and not on, say, type-theory, while others are fascinated by verification speed. None of these things were mentioned in the README and in my opinion other similar things of preference should also be omitted.

**Comment by benjub** [comment]
Typos:
Matamath --> Metamath (twice)
rigorously-verified --> rigorously verified

**Comment by david-a-wheeler** [comment]
@savask :

> I hope this doesn't come off as a personal grudge against multiple verifiers (because it's not), but my main message is that Metamath community members may have different views on things and disputable places should be avoided. Some, like Norman, value human-readability of set.mm and it being minimized almost to the byte-count, some like its focus on ZFC and not on, say, type-theory, while others are fascinated by verification speed. None of these things were mentioned in the README and in my opinion other similar things of preference should also be omitted.

It's okay and *normal* that different people find different things important. I see no problem there. However, the README should at least briefly mention *why* someone might use this repo instead of something else. All products represent trade-offs; new readers should be able to quickly learn about its properties to see if any of those properties are important to them. Some may not value readability, multiple verifiers, or verification speeds, but these properties are present & are important to some, so let's briefly tell them about it.

So getting back to the point: As you requested, I'll shorten the text about multiple verifiers in README.md as you suggested, and instead move those details to a different markdown file so that those who *do* care can learn more. I also plan to add a sentence or two about human readability, verification speed, and size, because it's important that a new reader learn about those properties. Thanks for noting those other important properties that should also have been mentioned!

> In the first section I would remove the phrase "Please join us to improve it further!", since it is a bit weird to ask for contributions before sections "What is Metamath?" and "How can I contribute?" :-) The penultimate section also asks for contributions, so a similar phrase can be removed from the first section, in my opinion.

I'll do that, but it's pretty common to invite contributions early on. We *want* people to contribute.

> In the third section I suggest removing everything before the phrase "The databases included and links to their generated displays, in (approximate) decreasing size, are:". The fact that a "database" is some sort of file is pretty obvious and becomes even clearer once the reader sees links to files in the database list.

It may be obvious to you, but the word "database" has implications to many non-Metamath people that are *not* true about Metamath databases. A lot of people think "database" means "SQL database" which isn't true but is a real problem. I'll remove that text in that section. But it's weird to use a word in a non-standard way without defining it. The text about Metamath (earlier) already uses the word "database", I think I can tweak the text there so that the word "database" is defined with only a few more words in the README.

> Also maybe I'm reading an older version, but I still see two occurrences of "Matamath" in the text.

Thanks! Fixed!

> ... mention only the bare minimum, something like:
> Changes ("commits") to any database are first automatically verified before they are accepted, using GitHub actions. Every change to set.mm and iset.mm is verified by several independent verifiers, most notably metamath-exe, which also checks markup, and mmj2, which checks definition soundness. All other databases' proofs are verified by one verifier (metamath-exe).

Ok, I'll do that. I'll change "most notably" to "including" because I'm not sure they're really the most notable, and I'll add a link from there to another file with more details if someone wants them (VERIFIERS.md). I also intend to keep the introductory first sentence to provide context.

Hopefully this gets us "close enough" :-).

**Review by benjub — APPROVED** [review]
There is still the hyphen in "rigorously-verified", but that can be fixed later,

**Comment by david-a-wheeler** [comment]
I believe the hyphen in "rigorously-verified" is optional in English (it doesn't matter either way). But since it doesn't matter either way, I'm happy to remove it. Removed.

**Comment by jkingdon** [comment]
> the word "database" has implications to many non-Metamath people that are _not_ true about Metamath databases. A lot of people think "database" means "SQL database"

Agreed that people would think first of SQL or something like redis, Dynamo, Mongo, etc. If I was writing this I'd probably try to figure out a way to avoid the word "database" entirely, but definitely that's not a project for this pull request and there's a much larger discussion about what a better word would be (if there is a better word).

**Comment by david-a-wheeler** [comment]
@jkingdon said:

> Agreed that people would think first of SQL or something like redis, Dynamo, Mongo, etc. If I was writing this I'd probably try to figure out a way to avoid the word "database" entirely, but definitely that's not a project for this pull request and there's a much larger discussion about what a better word would be (if there is a better word).

It's not obvious what a better word would be, though. I think we should just briefly clarify what *we* mean & move on. I slipped its definition in using just a few words in the explanation of what is Metamath, and I think that'll do.

**Comment by david-a-wheeler** [comment]
We have 5 approvals, and I tried to resolve the comments of @savask . I think this is overall an improvement from before. I intend to merge this in an hour or two. We can always make further improvements later, e.g., creating a separate file listing some other known Metamath databases as hinted at in #2384 .

**Comment by david-a-wheeler** [comment]
We have 5 approvals and I tried to resolve the comments of @savask . I'm merging, and we can make further improvements from there. I think this is *much* improved over current README which doesn't tell newcomers what the heck this is :-).

**Comment by savask** [comment]
@david-a-wheeler Thanks for taking my suggestions into account.
